File: problems/sorting/LC121_best_time_to_buy_and_sell_stock/solution.py
class Solution:
  def maxProfit(self, prices: List[int]) -> int:
    profit = 0
    buy_price = prices[0]
    for price in prices:
      if price > buy_price:
        profit = max(profit, price - buy_price)
      else:
        buy_price = price
    return profit
  
def LC121Main():
  nums = [7,1,5,3,6,4]
  s = Solution()
  profit = s.maxProfit(nums)
  print(profit)

LC121Main()

File: problems/sorting/LC121_best_time_to_buy_and_sell_stock/Solution.java
package sorting.LC121_best_time_to_buy_and_sell_stock;

public class Solution {
  public static int maxProfit(int[] prices) {
    int profit = 0;
    int buyPrice = prices[0];
    for (int price : prices) {
      if (price > buyPrice) {
        profit = Math.max(profit, price - buyPrice);
      } else {
        buyPrice = price;
      }
    }
    return profit;
  }
  public static void main(String[] args) {
    int[] prices = new int[]{7, 1, 5, 3, 6, 4};
    System.out.println(maxProfit(prices));
  }
}
